include/cuco/static_multimap.cuh

```diff
@@ -0,0 +1,784 @@
+ * - Host-side "bulk" operations
+ * - Device-side "singular" operations
+ *
+ * The host-side bulk operations include `insert`, `find`, and `contains`. These
+ * APIs should be used when there are a large number of keys to insert or lookup
+ * in the map. For example, given a range of keys specified by device-accessible
+ * iterators, the bulk `insert` function will insert all keys into the map.
+ *
+ * The singular device-side operations allow individual threads to perform
+ * independent insert or find/contains operations from device code. These
+ * operations are accessed through non-owning, trivially copyable "view" types:
+ * `device_view` and `mutable_device_view`. The `device_view` class is an
+ * immutable view that allows only non-modifying operations such as `find` or
+ * `contains`. The `mutable_device_view` class only allows `insert` operations.
+ * The two types are separate to prevent erroneous concurrent insert/find
+ * operations.
+ *
+ * Example:
+ * \code{.cpp}
+ * int empty_key_sentinel = -1;
+ * int empty_value_sentinel = -1;
+ *
+ * // Constructs a map with 100,000 slots using -1 and -1 as the empty key/value
+ * // sentinels. Note the capacity is chosen knowing we will insert 50,000 keys,
+ * // for an load factor of 50%.
+ * static_multimap<int, int> m{100'000, empty_key_sentinel, empty_value_sentinel};
+ *
+ * // Create a sequence of pairs {{0,0}, {1,1}, ... {i,i}}
+ * thrust::device_vector<thrust::pair<int,int>> pairs(50,000);
+ * thrust::transform(thrust::make_counting_iterator(0),
+ *                   thrust::make_counting_iterator(pairs.size()),
+ *                   pairs.begin(),
+ *                   []__device__(auto i){ return thrust::make_pair(i,i); };
+ *
+ *
+ * // Inserts all pairs into the map
+ * m.insert(pairs.begin(), pairs.end());
+ *
+ * // Get a `device_view` and passes it to a kernel where threads may perform
+ * // `find/contains` lookups
+ * kernel<<<...>>>(m.get_device_view());
+ * \endcode
+ *
+ *
+ * @tparam Key Arithmetic type used for key
+ * @tparam Value Type of the mapped values
+ * @tparam ProbeSequence Probe sequence used in multimap
+ * @tparam Scope The scope in which insert/find operations will be performed by
+ * individual threads.
+ * @tparam Allocator Type of allocator used for device storage
+ */
+template <typename Key,
+          typename Value,
+          class ProbeSequence      = cuco::double_hashing<Key, Value>,
+          cuda::thread_scope Scope = cuda::thread_scope_device,
+          typename Allocator       = cuco::cuda_allocator<char>>
+class static_multimap {
+  template <typename T>
+  static constexpr bool is_CAS_safe =
+    std::is_trivially_copyable_v<T>and std::has_unique_object_representations_v<T>;
+
+  static_assert(is_CAS_safe<Key>,
+                "Key type must be trivially copyable and have unique object representation.");
+  static_assert(is_CAS_safe<Value>,
+                "Value type must be trivially copyable and have unique object representation.");
```

Comment:
Can you change this based on the changes I made here? https://github.com/NVIDIA/cuCollections/pull/83 

The `is_bitwise_comparable` trait should get moved into it's own public header. Probably `traits.hpp`.

Reply:
Done